File: basics/go-programs/07_file_io.go
package main

import (
	"fmt"
	"os"
	"bufio"
)

/*
	GO PROGRAM 7: FILE I/O (READING & WRITING)
	Concepts:
	- 'os' package for file operations.
	- 'bufio' for buffered reading/writing.
	- 'defer' is used to ensure a cleanup operation (like closing a file) happens at the end of the function.
*/

func main() {
	filename := "testfile.txt"
	content := "Line 1: Go is awesome!\nLine 2: File handling is easy too."
	
	// WRITE FILE
	fmt.Printf("Writing to: %s\n", filename)
	
	err := os.WriteFile(filename, []byte(content), 0644)
	if err != nil {
		fmt.Printf("Error writing file: %v\n", err)
		return
	}
	
	// READ FILE
	fmt.Println("Reading from file:")
	
	file, err := os.Open(filename)
	if err != nil {
		fmt.Printf("Error opening file: %v\n", err)
		return
	}
	defer file.Close() // ALWAYS Close files when finished!
	
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		fmt.Printf("Line found: %s\n", scanner.Text())
	}
	
	if err := scanner.Err(); err != nil {
		fmt.Printf("Error during scan: %v\n", err)
	}
}

File: basics/node-programs/07_file_io.js
const fs = require('fs/promises');

/*
	NODE.JS PROGRAM 7: FILE I/O (READING & WRITING)
	Concepts:
	- 'fs/promises' for easy async file operations.
	- 'await fs.writeFile()' and 'await fs.readFile()'.
	- Converting data between strings and buffers.
*/

async function main() {
	const filename = "testfile_node.txt";
	const content = "Line 1: Node.js is powerful!\nLine 2: File handling is fully asynchronous.";
	
	// WRITE FILE
	console.log(`Writing to: ${filename}`);
	try {
		await fs.writeFile(filename, content);
		console.log("File written successfully!");
	} catch (err) {
		console.error("Error writing file:", err);
		return;
	}
	
	// READ FILE
	console.log("\nReading from file:");
	try {
		// utf8 encoding is important for reading text
		const data = await fs.readFile(filename, 'utf8');
		console.log("File Content Found:");
		console.log(data);
	} catch (err) {
		console.error("Error reading file:", err);
	}
}

main();
